### ISSUE TITLE
Some ideas for implementing max_workers

### ISSUE THREAD

**Issue by cjrh** [cjrh/coroexecutor#2]
Should be possible to implement either with a queue or a semaphore:

```python
import random
import asyncio

class xiter:
    def __init__(self, seq):
        self.seq = seq

    def __iter__(self):
        return iter(self.seq)

    def __aiter__(self):
        async def inner():
            for v in self.seq:
                yield v

        return inner()

async def amap(self, func, seq, *args, n=5, timeout=None):
    # Make a queue filled with tokens.
    q = asyncio.Queue(maxsize=n)
    for i in range(n):
        q.put_nowait(i)

    async def consumer(job_token, value):
        try:
            await asyncio.wait_for(
                func(value),
                timeout=timeout,
            )
        except asyncio.TimeoutError:
            pass
        finally:
            await q.put(job_token)

    async def producer(jobs):
        tasks = []
        async for job in jobs:
            # Is only allowed to run a job if there's a token available
            token = await q.get()
            # Consumer is responsible for putting a token back onto q
            # print('creating task')
            t = asyncio.create_task(consumer(token, job))
            tasks.append(t)
            # Remove done tasks from the list to prevent accumulation
            # Probably only want to do this every n jobs
            tasks = [t for t in tasks if not t.done()]

        print(f'Waiting for {len(tasks)} tasks')
        await asyncio.gather(*tasks)

    await producer(seq)
    print('work done')

async def amap2(self, func, seq, *args, n=5, timeout=None):
    sem = asyncio.Semaphore(n)

    async def consumer(job_token, value):
        async with sem:
            await asyncio.wait_for(
                func(value),
                timeout=timeout
            )

    async def producer(jobs):
        tasks = []
        async for job in jobs:
            t = asyncio.create_task(consumer(random.random(), job))
            tasks.append(t)
            # Remove done tasks from the list to prevent accumulation
            # Probably only want to do this every n jobs
            tasks = [t for t in tasks if not t.done()]

        print(f'Waiting for {len(tasks)} tasks')
        await asyncio.gather(*tasks)

    await producer(seq)
    print('work done')

await amap(f, xiter(range(100)), n=10)

await amap2(f, xiter(range(100)), n=10)
```

There is a difference between the two approaches. With the queue, tasks are only created as necessary. With the semaphore, all the tasks are created upfront, and the semphore (inside `consumer` controls the actual running of the jobs.  It may be possible to refactor this so that the task creation is also beneath the semaphpore. Haven't looked at that too closely yet.

**Comment by cjrh** [comment]
@kamadorueda I had a look at your suggestion, thanks for that. I started with very similar ideas around using a queue to "collect" submitted jobs, similar to my first comment at the top here.  I went through a bunch of permutations, but in the end I decided that it was all much simpler if I discarded the queue idea completely and simply allow consumers to have to `await executor.submit()` directly. This way back-pressure is applied in a very clear and transparent way. I'm quite happy with the implementation now in v2021.5.1 which has been published on PyPI.

Thanks again for your interest :)

Closing this issue.
